modules/EvseManager/Charger.cpp

```diff
@@ -1167,31 +1167,131 @@ bool Charger::deauthorize_internal() {
     return false;
 }
 
-bool Charger::disable(int connector_id) {
+bool Charger::enable_disable(int connector_id, const types::evse_manager::EnableDisableSource& source) {
     Everest::scoped_lock_timeout lock(state_machine_mutex, Everest::MutexDescription::Charger_disable);
-    if (connector_id not_eq 0) {
-        shared_context.connector_enabled = false;
+
+    // insert the new request into the table
+    bool replaced = false;
+    for (auto& entry : enable_disable_source_table) {
+        if (entry.enable_source == source.enable_source) {
+            // replace this entry as it is an update from the same source
+            entry = source;
+            replaced = true;
+        }
     }
-    shared_context.current_state = EvseState::Disabled;
-    signal_simple_event(types::evse_manager::SessionEventEnum::Disabled);
-    return true;
-}
 
-bool Charger::enable(int connector_id) {
-    Everest::scoped_lock_timeout lock(state_machine_mutex, Everest::MutexDescription::Charger_enable);
+    if (not replaced) {
+        // Add to the table
+        enable_disable_source_table.push_back(source);
+    }
+
+    // Find out if we are actually enabled or disabled at the moment
+    bool is_enabled = parse_enable_disable_source_table();
+
+    if (is_enabled) {
+        if (connector_id not_eq 0) {
```

Comment:
Shouldn't there be a check that the connector ID does correspond to this charger's connector ID?

Reply:
The EvseManager should only get calls for connector_id(s) that it owns. This is a bit confusing at the moment since right now EvseManager supports only one connector per Evse. OCPP2 already supports multiple connectors per evse. EvseManager will need to change architecture a bit, then this will be cleaner